docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,502 @@
+- TBD
+
+creation-date: 2022-01-05
+
+last-updated: 2022-01-05
+
+---
+
+# Parameterized Config Management Plugins
+
+Config Management Plugin (CMP) parameterization defines a way for plugins to "announce" and then consume acceptable 
+parameters for an Application. Announcing parameters allows CMPs to provide a UI experience on par with native config 
+management tools (Helm, Kustomize, etc.).
+
+- [Parameterized Config Management Plugins](#parameterized-config-management-plugins)
+  * [Open Questions](#open-questions)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    + [1. CMPs are under-utilized](#1-cmps-are-under-utilized)
+    + [2. Decisions about config management tools are limited by the core code](#2-decisions-about-config-management-tools-are-limited-by-the-core-code)
+    + [3. Ksonnet is deprecated, and CMPs are a good place to maintain support](#3-ksonnet-is-deprecated-and-cmps-are-a-good-place-to-maintain-support)
+    + [Goals](#goals)
+    + [Non-Goals](#non-goals)
+  * [Proposal](#proposal)
+    + [Use cases](#use-cases)
+      - [Use case 1: building Argo CD without config management dependencies](#use-case-1-building-argo-cd-without-config-management-dependencies)
+      - [Use case 2: writing CMPs with rich UI experiences](#use-case-2-writing-cmps-with-rich-ui-experiences)
+    + [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      - [Prerequisites](#prerequisites)
+      - [Terms](#terms)
+      - [CMP config schema](#cmp-config-schema)
+      - [Parameters manifest / parameters serialization format](#parameters-manifest--parameters-serialization-format)
+      - [Parameter definition schema](#parameter-definition-schema)
+      - [Parameters announcement schema](#parameters-announcement-schema)
+      - [Parameter list schema](#parameter-list-schema)
+    + [Detailed examples](#detailed-examples)
+      - [Example 1: trivial parameterized CMP](#example-1-trivial-parameterized-cmp)
+      - [Example 2: Helm parameters from Kustomize dependency](#example-2-helm-parameters-from-kustomize-dependency)
+      - [Example 3: simple Helm CMP](#example-3-simple-helm-cmp)
+    + [Security Considerations](#security-considerations)
+      - [Increased scripting](#increased-scripting)
+    + [Risks and Mitigations](#risks-and-mitigations)
+    + [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  * [Drawbacks](#drawbacks)
+  * [Alternatives](#alternatives)
+
+## Open Questions
+
+Is there a way to make it easier for the generate command to consume parameters? (Easier than parsing a JSON object, 
+that is.)
+* We could translate all parameters to env vars, like `ARGOCD_PARAM_{section}_{param name}`.
+  * Advantage: no need to pull the param from JSON.
+  * Disadvantage: we'd need an easy-to-predict and collision-resistant way to translate arbitrary parameter names
+    to env vars. 
+    * Option 1: Enforce some parameter name convention which allows a 1:1 translation to env
```

Comment:
For option 1, the plugin might not have control over the parameter names.  For example, the parameters for the helm built-in come from whatever the application creator puts in the values file.